File: concepts/04_collections_and_iteration/03_sequence_mutation_and_invalidation/lua/test_02_views_and_structural_changes.lua
-- Common question: do subcollections alias storage, and what invalidates a view?
-- Inputs: table aliases, table.move copies, nested references, and source mutation.
-- Observations: alias visibility, independent top-level structure, shallow-copy sharing, and no native views.
-- polyglot-family: collections_and_iteration
-- polyglot-concept: sequence_mutation_and_invalidation
-- polyglot-related: languages/lua/language/01_values_and_types/test_007_reference_identity.lua

local t = require("support.assertions")
local source = {{value = 1}, {value = 2}, {value = 3}}
local alias = source
local copied = {}
table.move(source, 1, 2, 1, copied)

alias[1] = {value = 10}
t.equal(source[1].value, 10)
t.equal(copied[1].value, 1)

copied[2].value = 20
t.equal(source[2].value, 20)

table.remove(source, 1)
t.equal(#source, 2)
t.equal(#copied, 2)
t.falsey(rawequal(source, copied))

-- Lua table 没有内建 slice view；复制的是引用值，嵌套对象仍可共享。
t.done()

File: concepts/04_collections_and_iteration/03_sequence_mutation_and_invalidation/ruby/test_02_views_and_structural_changes.rb
# 共同问题：子集合是否共享底层存储，结构变化会不会令 view 失效。
# 输入：Array alias、slice、嵌套对象和 source 突变；观察：Ruby slice 为浅复制而非原生 view。
# polyglot-family: collections_and_iteration
# polyglot-concept: sequence_mutation_and_invalidation
# polyglot-related: languages/ruby/language/06_collections_and_enumeration/
# polyglot-related+: test_041_array_indexing_slicing_and_mutation.rb

require "assertions"

A = PolyglotAssertions

nested = +"shared"
source = [nested, "middle", "end"]
alias_value = source
slice = source[0, 2]
alias_value << "new"
A.equal(4, source.length)
A.equal(2, slice.length)
A.falsey(slice.equal?(source))
A.same(nested, slice.first)

nested << "-changed"
A.equal("shared-changed", source.first)
A.equal("shared-changed", slice.first)
source.delete_at(1)
A.equal(["shared-changed", "middle"], slice)

A.done
